### PR TITLE
Improve inversion in f64 field

### PR DESCRIPTION
**Pull request by irakliyk**
This PR improves inversion in the `f64` field by replacing the current implementation with the method based on Fermat’s little theorem. Specifically, we fully unroll the multiplication by `M - 2` which allows us to achieve modular in version in 72 multiplications.

The net result is that this implementation is about 2.5x faster than the previous one.

### CODE REVIEW

File: math/src/field/f64/mod.rs [hunk 1]
```diff
@@ -99,51 +99,33 @@ impl FieldElement for BaseElement {
...
+        let t12 = exp_acc::<6>(t6, t6);
 
-        while a > (M as u128) {
-            a -= M as u128;
-        }
+        // compute base^111111111111111111111111
+        let t24 = exp_acc::<12>(t12, t12);
 
-        Self(a as u64)
+        // compute base^111111111111111111111111
```
Comment (kchalkias) [inline math/src/field/f64/mod.rs:120]: replace with // compute base^111111111111111111111111111111 (30 ones)

**Review by kchalkias — APPROVED** [review]

File: math/src/field/f64/mod.rs [hunk 2]
```diff
@@ -99,51 +99,33 @@ impl FieldElement for BaseElement {
     #[inline]
     #[allow(clippy::many_single_char_names)]
     fn inv(self) -> Self {
-        let x = self.as_int();
+        // compute base^(M - 2) using 72 multiplications
+        // M - 2 = 0b1111111111111111111111111111111011111111111111111111111111111111
```
Comment (dbeaverfb) [inline math/src/field/f64/mod.rs:104]: // t_i represents base^{2^i}

File: math/src/field/f64/mod.rs [hunk 3]
```diff
@@ -569,3 +551,13 @@ fn mod_reduce(x: u128) -> u64 {
     let (result, over) = tmp0.overflowing_add(tmp1);
     result.wrapping_add(E * (over as u64))
 }
+
+/// Squares the base N number of times and multiplies the result by the tail value.
```
Comment (dbeaverfb) [inline math/src/field/f64/mod.rs:555]: So: logarithmic terms:      (logx << N) + logy.

(No need to use this as a comment, just confirming.)

File: math/src/field/f64/mod.rs [hunk 4]
```diff
@@ -569,3 +551,13 @@ fn mod_reduce(x: u128) -> u64 {
...
+#[inline(always)]
+fn exp_acc<const N: usize>(base: BaseElement, tail: BaseElement) -> BaseElement {
+    let mut result = base;
+    for _ in 0..N {
+        result = result.square();
+    }
+    result * tail
+}
```
Comment (dbeaverfb) [inline math/src/field/f64/mod.rs:563]: This is very nicely written and optimized.

I only see unit tests for 0 and 1.  Maybe I am missing something.  To know that this isn't a regression, maybe a unit test with 2 or some other larger number would help.  Even using the values calculated from the pre-existing code would be fine.

**Review by dbeaverfb — APPROVED** [review]

File: math/src/field/f64/mod.rs (hunk 1 above)
Comment (bobbinth) [reply math/src/field/f64/mod.rs]: fixed!

File: math/src/field/f64/mod.rs (hunk 2 above)
Comment (bobbinth) [reply math/src/field/f64/mod.rs:104]: kind of: up until `t31` that's how it works. beyond that, the number is just for the number of bits of the exponent included in the temp value.

File: math/src/field/f64/mod.rs [hunk 5]
```diff
@@ -569,3 +551,13 @@ fn mod_reduce(x: u128) -> u64 {
     let (result, over) = tmp0.overflowing_add(tmp1);
     result.wrapping_add(E * (over as u64))
 }
+
+/// Squares the base N number of times and multiplies the result by the tail value.
+#[inline(always)]
+fn exp_acc<const N: usize>(base: BaseElement, tail: BaseElement) -> BaseElement {
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply math/src/field/f64/mod.rs:557]: I was going for something like "exponent_accumulate" :)

File: math/src/field/f64/mod.rs (hunk 4 above)
Comment (bobbinth) [reply math/src/field/f64/mod.rs:563]: There are randomized tests in `inv_proptest()` which should catch issues with other values.

File: math/src/field/f64/mod.rs (hunk 1 above)
Comment (irakliyk) [reply math/src/field/f64/mod.rs]: Fixed!

File: math/src/field/f64/mod.rs (hunk 2 above)
Comment (irakliyk) [reply math/src/field/f64/mod.rs:104]: kind of: up until `t31` that's how it works. beyond that, the number is just for the number of bits of the exponent included in the temp value.

File: math/src/field/f64/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (irakliyk) [reply math/src/field/f64/mod.rs:557]: I was going for something like "exponent_accumulate" :)

File: math/src/field/f64/mod.rs (hunk 4 above)
Comment (irakliyk) [reply math/src/field/f64/mod.rs:563]: There are randomized tests in `inv_proptest()` which should cover all other values.

File: math/src/field/f64/mod.rs (hunk 3 above)
Comment (irakliyk) [reply math/src/field/f64/mod.rs:555]: I think so.

**Pull request merged**